posthog/temporal/llm_analytics/trace_clustering/data.py

```diff
@@ -0,0 +1,142 @@
+"""Data access layer for trace clustering.
+
+This module consolidates all ClickHouse queries used by the clustering workflow,
+providing a single source of truth for data fetching operations.
+"""
+
+from datetime import datetime
+from typing import Any
+
+from posthog.clickhouse.client.connection import Workload
+from posthog.clickhouse.client.execute import sync_execute
+from posthog.temporal.llm_analytics.trace_clustering import constants
+from posthog.temporal.llm_analytics.trace_clustering.models import (
+    TraceEmbeddingRow,
+    TraceEmbeddings,
+    TraceId,
+    TraceSummaries,
+    TraceSummaryRow,
+)
+
+
+def _rows_to_dicts(rows: list[tuple[Any, ...]], column_types: list[tuple[str, str]]) -> list[dict[str, Any]]:
+    """Convert ClickHouse rows to dictionaries using column names.
+
+    This provides safe column access by name rather than fragile index-based access.
+    The returned dicts should be cast to appropriate TypedDicts by callers for type safety.
+    """
+    column_names = [col[0] for col in column_types]
+    return [dict(zip(column_names, row)) for row in rows]
+
+
+def fetch_trace_embeddings_for_clustering(
+    team_id: int,
+    window_start: datetime,
+    window_end: datetime,
+    max_samples: int,
+) -> tuple[list[TraceId], TraceEmbeddings]:
+    """Query trace IDs and embeddings from document_embeddings table.
+
+    Args:
+        team_id: Team ID to query embeddings for
+        window_start: Start of time window
+        window_end: End of time window
+        max_samples: Maximum number of traces to sample
+
+    Returns:
+        Tuple of (list of trace IDs, dict mapping trace_id -> embedding vector)
+    """
+    query = """
+        SELECT document_id, embedding
+        FROM distributed_posthog_document_embeddings
+        WHERE team_id = %(team_id)s
+            AND timestamp >= %(start_dt)s
+            AND timestamp < %(end_dt)s
+            AND product = %(product)s
+            AND document_type = %(document_type)s
+            AND rendering = %(rendering)s
+            AND length(embedding) > 0
+        ORDER BY rand()
+        LIMIT %(max_samples)s
+    """
+    rows, column_types = sync_execute(
+        query,
+        {
+            "team_id": team_id,
+            "start_dt": window_start,
+            "end_dt": window_end,
+            "product": constants.LLMA_TRACE_PRODUCT,
+            "document_type": constants.LLMA_TRACE_DOCUMENT_TYPE,
+            "rendering": constants.LLMA_TRACE_DETAILED_RENDERING,
+            "max_samples": max_samples,
+        },
+        workload=Workload.OFFLINE,
+        with_column_types=True,
+    )
+
+    results: list[TraceEmbeddingRow] = _rows_to_dicts(rows, column_types)  # type: ignore[assignment]
+    trace_ids = [row["document_id"] for row in results]
+    embeddings_map = {row["document_id"]: row["embedding"] for row in results}
```

Comment:
Paranoid: Build the map in the same loop to ensure ids and embeddings are 100% in the same order?

Reply:
cleaned that up here - https://github.com/PostHog/posthog/pull/41954/commits/3e6ae09c7353ee674cee4c684d96a269ae70c410

(was actually looking around if any hogql to pandas df as data here not huge and could make code simpler but not sure worth it if not already established pattern)